### PR TITLE
[Core] MLS shape functions enhancements

### PR DESCRIPTION
**Pull request by rubenzorrilla**
**📝 Description**
In this PR I'm enhancing the current MLS shape functions capabilities. The main improvements are:
- Avoiding the transpose matrix multiplication in the L2 minimization problem. Instead a QR decomposition is used to solve the rectangular matrix. This should make the resolution of the problem more robust in presence of ill-conditioned cases.
- Adding the capability of using a 2nd order interpolation basis.
- Kernel function simplification (no need to normalize with Pi).

Note that these required quite a lot of refactoring (this is why there are so much changes).

@frawahl FYI, this are the changes I told you long time ago.

### CODE REVIEW

File: kratos/utilities/mls_shape_functions_utility.h [hunk 1]
```diff
@@ -37,6 +38,8 @@ class MLSShapeFunctionsUtility
 
 public:
 
+    using DenseSpace = UblasSpace<double, Matrix, Vector>;
```
Comment (philbucher) [inline kratos/utilities/mls_shape_functions_utility.h]: Seems unused here
Maybe move to cpp?

Or did you mean to use the ublas interface instead?

Comment (rubenzorrilla) [reply kratos/utilities/mls_shape_functions_utility.h]: I think I added it when trying different QR decompositions. Removing it.

Comment (rubenzorrilla) [reply kratos/utilities/mls_shape_functions_utility.h]: OFC, it is used in the cpp to instantiate the QR :facepalm:

**Review by RiccardoRossi — COMMENTED** [review]
why are u changing from stack matrices and vectors to tje dynqmoc versions?

**Comment by rubenzorrilla** [comment]
> why are u changing from stack matrices and vectors to tje dynqmoc versions?



> why are u changing from stack matrices and vectors to tje dynqmoc versions?

Note that we're changing the way we calculate the approximants (no transpose multiplication. Now it is impossible as the problem size is dependent on the  number of points in the cloud.

File: kratos/utilities/mls_shape_functions_utility.h [hunk 2]
```diff
@@ -23,6 +23,7 @@
 // Project includes
 #include "containers/array_1d.h"
 #include "includes/define.h"
+#include "spaces/ublas_space.h"
```
Comment (philbucher) [inline kratos/utilities/mls_shape_functions_utility.h]:
```suggestion
#include "includes/define.h"
#include "includes/ublas_interface.h"
```

the ublas_space should be moved to the cpp

Comment (rubenzorrilla) [reply kratos/utilities/mls_shape_functions_utility.h]: Done.

**Review by RiccardoRossi — APPROVED** [review]
9k to me

**Pull request merged**